lib/kafka_consumer/offset_server.ex

```diff
@@ -0,0 +1,145 @@
+defmodule KafkaConsumer.OffsetServer do
+  @type offset :: number
+  @type parition :: number
+  @type noreply :: {:noreply, offset}
+  @type reply :: {:reply, offset, offset}
+  @type topic :: String.t
+
+  @callback increment(topic) :: noreply
+  @callback decrement(topic) :: noreply
+  @callback get(topic) :: reply
+  @callback save(topic) :: reply
+  @callback start_link(topic, parition) :: {:ok, pid}
+
+
+  defmacro __using__(_) do
+    quote do
+      @type offset :: number
+      @type partition :: number
+      @type noreply :: {:noreply, offset}
+      @type reply :: {:reply, offset, offset}
+      @type topic :: String.t
+
+      @moduledoc """
+      Holds the last processed offset
+      All servers are registerd with an atom of the topic
+
+      `start_link("that.one.topic")` will register a server under `:that_one_topic`
+      """
+      use GenServer
+
+      @behaviour KafkaConsumer.OffsetServer
+
+      @doc """
+      Starts a OffsetServer for the given topic
+      """
+      @spec start_link(topic, partition) :: {:ok, pid}
+      def start_link(topic, partition) do
+        name = make_server_name(topic)
+        GenServer.start_link(
+          __MODULE__,
+          get_latest_offset(topic, partition),
+          name: name,
+          id: name
+        )
+      end
+
+      @doc """
+      increments the offset
+      """
+      @spec increment(topic) :: noreply
+      def increment(topic) do
+        GenServer.cast(make_server_name(topic), :increment)
+      end
+
+      @doc """
+      decrements the offset
+      """
+      @spec decrement(topic) :: noreply
+      def decrement(topic) do
+        GenServer.cast(make_server_name(topic), :decrement)
+      end
+
+      @doc """
+      saves the offset
+      """
+      @spec increment(topic) :: noreply
+      def save(topic) do
+        GenServer.cast(make_server_name(topic), :save)
+      end
```

Comment:
I'd argue that the user should implement That logic, as different users
would like different things.

Some might want an ETS table, while others want to store the offset in a
database.

Thoughts?

On 12 Sep 2016 9:14 pm, "Andrew Noskov" <notifications@github.com> wrote:

> In lib/kafka_consumer/offset_server.ex
> <https://github.com/anoskov/kafka-consumer/pull/6#discussion_r78353615>:
>
> > +
> > +      @doc """
> > +      decrements the offset
> > +      """
> > +      @spec decrement(topic) :: noreply
> > +      def decrement(topic) do
> > +        GenServer.cast(make_server_name(topic), :decrement)
> > +      end
> > +
> > +      @doc """
> > +      saves the offset
> > +      """
> > +      @spec increment(topic) :: noreply
> > +      def save(topic) do
> > +        GenServer.cast(make_server_name(topic), :save)
> > +      end
>
> I think it must be fault tolerant. You can try store and get it from DTS
> table as example. https://github.com/zackehh/stash
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/anoskov/kafka-consumer/pull/6/files/a4689ffc462088bdeb8673c199a0990b13417013#r78353615>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5r2OW5tcGv0UASj8YRNEyx3NYQdNPrks5qpTQPgaJpZM4J5qti>
> .
>